packages/wiki-docs/gimp/GIMP-1-i18n.md

```diff
@@ -0,0 +1,119 @@
+* namespaces.json `import` will be set to `['gp', 'gp-testo']` namespaces. This will ensure non-en languages get imported on every import or sync for both core and project-specific translations.
+* namespaces.json `publish` will be left at `['gp-testo']` so that core translations are not published back to the main project.
+* As developer edits reports/components that were installed on init from template, they should maintain the `gp` namespace, as long as no part of the string to translate changes.  If it changes, then the namespace should switch to the project namespace (e.g. `gp-testo`).  On the next `translation:extract` the string will then get moved to the new namespace, and on the next `translation:import` it will get published to the `gp-testo` tag in POEDITOR to be re-translated for all languages, and then translations for other languages will get imported back to the project on next `translation:import`.
+  * Why not switch the whole component (all translated strings) from the `gp` namespace over to the project namespace if one string gets changed?  Because that will force extra re-translation.  (may reconsider).
+  * Failure to switch away from the `gp` namespace means your string won't get published to poeditor for translation, because the `gp` namespace is not a published namespace for this use case.
+  * There may need to be multiple useTranslation(ns) calls within a single component, if there are strings for more than one namespace.  Rename the t function as needed to differentiate.
+* Translator will use POEDITOR and they will filter by `gp` tag for core translations, and by `gp-testo` tag for project translations.
+
+### 2. Reports developed by external developer using POEDITOR
+
+* The user will create their own POEDITOR project and get the environment variables for project and api key setup.
+* If POEditor free account does not suffice, then if you make the report code public and open source, then you can request upgrade to a free premium account in POEditor interface.
+* namespaces.json `import` will be set to `['gp', 'gp-testo']` namespaces.  This will ensure updated translations for non-en languages get imported on every import or sync for both core and project-specific translations.
+* namespaces.json `publish` will be set to `['gp', 'gp-testo']` so that core and project translations are published to the new POEDITOR project making it the source of truth.  There will be no connection to the SSN POEDITOR project.
+* As developer edits reports/components that were installed on init from template, they should maintain the `gp` namespace for strings that already had it for consistency.  This will ensure existing translations for other languages are not lost.
+  * In this situation, there may need to be multiple useTranslation(ns) calls within a single component, if there are strings for more than one namespace.  Rename the t function as needed to differentiate.
```

Comment:
Dealing with namespaces sounds pretty confusing, and I'm not sure I quite understand myself how this works at this point. Is there some way to simplify this for the report author so that they can just use usTranslation and t("") and not have to worry about it? My understanding is that the gp namespace strings will be embedded in core components so the author doesn't really need to explicitly reference anything in the gp namespace. 

Could you provide a wrapper for <Trans> and a replacement useTranslation that automatically assigns strings to a project namespace? Or maybe that's even overly complicated, and project strings could be _un-namespaced_ while core strings are in gp? Maybe that wouldn't be compatible with the namespaces.json publishing scheme...

Reply:
The gp namespace is used for core components, but it's also used for template report clients and components that the developer will install into their project on init so they will be exposed to the gp namespace and as they customize them they'll need to grapple with whether to change the namespace from gp to their project namespaces name.

Reply:
Sorry made that last comment before seeing your reply pop up, which clarifies things. I think if the templates generate Trans tags with ns clearly specified it will probably be easy to get what's going on. Maybe if usTranslation is being used a lot in templates it could be setup to generate something like this:

```ts

const {t} = useTranslation("myProject");

return <div>
  <h1>{t("eezReportTitle", {ns: "gp"})}</h1>
</div>
```

Then if people start adding text to their project they can use `t` without too much thinking to start.

Reply:
Interesting idea to not namespace and add it on publish to POEditor.  It would have to work all the way through though.

That would mean both template and project strings would both not be namespaced.

The hiccup there is we are pre-translating all the template reports, along with the core UI components.  Would need to keep those straight in POEditor as well.  That's why they get the `gp` namespace assigned too.  Having no namespace for template translated strings would mean they get dumped into the default `translated` namespace in the SSN POEditor project, which I want to avoid.

Maybe this would make a good reason to just have a separate geoprocessing poeditor project, separate from ssn.  Then we can still have project specific strings auto-tagged by the publish script.  Here's the relevant snippet in publishTerms.ts

```javascript
        termsToAdd.push({
          term: localTermKey,
          english: localTerms[localTermKey],
          tags: [namespace],
        });
```

Reply:
@underbluewaters I'm not sure what you mean by "templates generate Trans tags with ns clearly specified".

Keep in mind that when you install a template on project init, it doesn't generate any template code, it just is a pre-developed report client, that's pre-translated, and it gets copied into the users project space.  That's it.

Here's what you get out of the box right now in this PR for the ocean eez templates SizeCard.

```jsx
const SizeCard = () => {
  const { t } = useTranslation("gp");
  return (
    <>
      <ResultsCard
        title={t("SizeCardTitle", "Zone Size")}
        functionName="calculateArea"
        skeleton={<LoadingSkeleton />}
      >
        {(data: AreaResults) => {
          return (
            <p>
              📐
              <Trans ns="gp" i18nKey="SizeCardAreaMsg">
                This sketch is{" "}
                <b>{{ area: Number.format(Math.round(data.area * 1e-6)) }}</b>{" "}
                square kilometers
              </Trans>
            </p>
          );
        }}
      </ResultsCard>
    </>
  );
};
```

If the report developer edits/changes the strings here, it's suggested in some use cases they need to change the namespace here to their project specific namespace.

I think it's worth a direct discussion at this point in order for there to be full understanding.